tools/pubsys/src/repo/fetch_variant/mod.rs

```diff
@@ -0,0 +1,240 @@
+use std::path::{Path, PathBuf};
+use tough::{Prefix, Repository, RepositoryLoader, TargetName};
+use url::Url;
+
+use buildsys::manifest::{ImageFormat, ManifestInfo, PartitionPlan};
+
+/// fetching and downdloaing the image targets of a given variant
+#[derive(Debug, Parser)]
+pub(crate) struct FetchVariantArgs {
+    #[arg(long)]
+    /// Use this named repo infrastructure from Infra.toml
+    repo: String,
+
+    #[arg(long)]
+    /// The architecture of the repo being validated
+    arch: String,
+
+    #[arg(long)]
+    /// The variant of the repo being validated
+    variant: String,
+
+    #[arg(long)]
+    /// Path to root.json for this repo
+    root_role_path: PathBuf,
+
+    #[arg(long)]
+    /// Where to store the downloaded img files
+    outdir: PathBuf,
+
+    #[arg(long)]
+    /// The variant name without extension
+    filename_prefix: String,
+
+    #[arg(long)]
+    /// The root directory of the buildsys
+    variant_manifest: PathBuf,
+}
+
+/// Download targets
+async fn handle_download(
+    repository: &Repository,
+    outdir: &Path,
+    raw_names: &[String],
+) -> Result<(), Error> {
+    let target_names: Result<Vec<TargetName>, Error> = raw_names
+        .iter()
+        .map(|s| TargetName::new(s).context(error::InvalidTargetNameSnafu))
+        .collect();
+    let target_names = target_names?;
+    let download_target = |name: TargetName| async move {
+        println!("\t-> {}", name.raw());
+        repository
+            .save_target(&name, outdir, Prefix::None)
+            .await
+            .context(error::MetadataSnafu)
+    };
+
+    // copy requested targets, or all available targets if not specified
+    let targets: Vec<TargetName> = if target_names.is_empty() {
+        repository
+            .targets()
+            .signed
+            .targets
+            .keys()
+            .cloned()
+            .collect()
+    } else {
+        target_names
+    };
+
+    println!("Downloading targets to {outdir:?}");
+    tokio::fs::create_dir_all(outdir)
+        .await
+        .context(error::DirCreateSnafu { path: outdir })?;
+    for target in targets {
+        download_target(target).await?;
+    }
+    Ok(())
+}
+
+/// Fetch the variant and download its image targets
+async fn fetch_variant(
+    root_role_path: &PathBuf,
+    metadata_url: Url,
+    targets_url: &Url,
+    outdir: &PathBuf,
+    filename_prefix: &str,
+    variant_manifest: &PathBuf,
+    variant: &str,
+) -> Result<(), Error> {
+    // Load the repository
+    let repo = RepositoryLoader::new(
+        &repo::root_bytes(root_role_path).await?,
+        metadata_url.clone(),
+        targets_url.clone(),
+    )
+    .load()
+    .await
+    .context(repo_error::RepoLoadSnafu {
+        metadata_base_url: metadata_url.clone(),
+    })?;
+
+    // To help ensure that downloads are safe, we require that the outdir does not exist.
+    ensure!(
+        !outdir.exists(),
+        error::DownloadOutdirExistsSnafu { path: outdir }
+    );
```

Comment:
I would recommend downloading to a temporary directory that you create, then copying the artifacts that match the expected prefix to the final output directory.

That will make it possible to re-run this step repeatedly and get the same result, rather than having to manually clean up directories between invocations.

Reply:
I don't mind writing to a temporary directory first, but we might want to keep this check to avoid the situation where someone fetches a variant that they already built locally (unless we want to clean up the directory as part of the fetch).